### DISCUSSION TITLE
Instancing multiple objects from list of parameter dictionaries

### DISCUSSION THREAD

**Discussion by joeylitalien** [mitsuba-renderer/mitsuba3#1009] [Q&A] [answered]
Hi!

I'm wondering if there is a way to instantiate objects from an array of parameter dictionaries directly in Python. For instance, I can create a single rough plastic BRDF from a dictionary with one roughness value, but I would like to do this for e.g. _N_ different roughness values in parallel without having to loop over my dictionaries. This is essentially what's done under the hood when tracing rays in a rendering loop but I'm looking for a way to achieve this without any scene. My use case is that I want to be able to query some of the BRDF class functions (e.g., `eval` and `pdf`) for arbitrary parameters in a batched manner for ML applications.

Thanks!

**Comment by njroussel** [comment, marked as answer]
Hi @joeylitalien !

Without going too much into the technical details, there is an implicit limitation of plugins that their members must be scalar (this is not exactly true, but for the purposes of this discussion we can consider that it is). Basically, a `BSDF` plugin can't have many roughness values, for example. You therefore need to instantiate one plugin per parameter configuration.

-------
**Suggestion 1** 

To improve the loading time, you could create a custom `BSDF` class, which just acts as a holder of the N `BSDF` you'd want, so you could declare a single massive dictionary:
```python
my_huge_dict = {
    'type': 'my_bsdf_holder',
    'bsdf1': {
        'type': 'principled',
        ....
    },
    'bsdf2': {
        'type': 'principled',
        ....
    },
    .....
}
```
The advantage of this, is that when you call `mi.load_dict(my_huge_dict)` you'd benefit from the parallel construction of the N inner plugins. You'd still need to define some accessor to fetch the inner BSDF from the custom one, but that can be done with a simple python class member.

Finally, you need to call all the BSDFs in a vectorized manner, you can see [this thread](https://github.com/mitsuba-renderer/mitsuba3/discussions/889) on how to build a `BSDFPtr` to do exactly that.

----------
**Suggestion 2** 

I can't imagine the first suggestion scaling very well with millions of instances. It would take a significant amount of time for Dr.Jit to record the `eval` method of each instance. 

If you really only need two methods from a specific BSDF plugin, I think it would be fairly easy to copy/re-write those functions in Python and replace the access to the scalar class members by some function argument which is the array of parameters you want to us. This has the following benefits:
* No plugins are loaded
* No virtual functions, Dr.Jit only needs to record that single new function
